File: backends/java-spring/src/main/java/com/authorization/infrastructure/cache/RedisInvitationAuthRepository.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.infrastructure.cache;

import com.authorization.config.AppConfig;
import com.authorization.domain.invitation.repositories.InvitationAuthRepository;
import org.springframework.stereotype.Component;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;

/**
 * Redis を使用した招待認証キャッシュRepositoryの実装です。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
@Component
public class RedisInvitationAuthRepository implements InvitationAuthRepository {

    private final JedisPool pool;
    private final String cachePrefix;

    /**
     * コンストラクタ。
     *
     * @param pool Redis接続プール
     * @param cfg アプリケーション設定
     */
    public RedisInvitationAuthRepository(JedisPool pool, AppConfig cfg) {
        this.pool = pool;
        this.cachePrefix = cfg.app().cachePrefix();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void store(String token, int role, int ttl) {
        try (Jedis jedis = pool.getResource()) {
            jedis.setex(cacheKey(token), ttl, String.valueOf(role));
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Integer find(String token) {
        try (Jedis jedis = pool.getResource()) {
            String v = jedis.get(cacheKey(token));
            if (v == null) {
                return null;
            }
            try {
                return Integer.parseInt(v);
            } catch (NumberFormatException e) {
                return null;
            }
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Integer consume(String token) {
        try (Jedis jedis = pool.getResource()) {
            String v = jedis.getDel(cacheKey(token));
            if (v == null) {
                return null;
            }
            try {
                return Integer.parseInt(v);
            } catch (NumberFormatException e) {
                return null;
            }
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void remove(String token) {
        try (Jedis jedis = pool.getResource()) {
            jedis.del(cacheKey(token));
        }
    }

    /**
     * キャッシュキーを組み立てます。
     *
     * @param token 招待トークン
     * @return キャッシュキー
     */
    private String cacheKey(String token) {
        return cachePrefix + ":invitation_auth:invitation_auth:" + token;
    }
}

File: backends/kotlin-ktor/src/main/kotlin/com/authorization/infrastructure/cache/RedisInvitationAuthRepository.kt
/*
 * 招待認証キャッシュリポジトリの Redis 実装モジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.infrastructure.cache

import com.authorization.config.Config
import com.authorization.domain.invitation.AuthRepository
import redis.clients.jedis.JedisPool

/**
 * Redis を使用した招待認証キャッシュリポジトリの実装です。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
class RedisInvitationAuthRepository(
    private val pool: JedisPool,
    private val cfg: Config,
) : AuthRepository {

    override suspend fun store(token: String, ttl: Long) {
        pool.resource.use { it.setex(cacheKey(token), ttl, token) }
    }

    override suspend fun find(token: String): String? =
        pool.resource.use { it.get(cacheKey(token)) }

    override suspend fun remove(token: String) {
        pool.resource.use { it.del(cacheKey(token)) }
    }

    private fun cacheKey(token: String): String =
        "${cfg.app.cachePrefix}:invitation_auth:invitation_auth:$token"
}
